Fix glfw3 default hints being modified

ypujante opened the pull request:
The code is doing: `GLFW.hints = GLFW.defaultHints` which is assigning the object `GLFW.defaultHints` to the `GLFW.hints` and **not** copying its value. As a result setting any hint after the fact (using `glfwWindowHint`) modifies the `GLFW.defaultHints` and as a result `glfwDefaultWindowHints` is not restoring the defaults.

The purpose of this PR is to fix this problem. I also added a test for it.

sbc100 commented on src/library_glfw.js:
```diff
@@ -996,6 +996,13 @@ var LibraryGLFW = {
+      GLFW.hints = {};
+      for (var k in GLFW.defaultHints) {
+        GLFW.hints[k] = GLFW.defaultHints[k];
+      }
```
I think you can write this as:

``
GLFW.hints = Object.assign({}, GLFW.defaultHints)
```

ypujante replied:
Yes that works and is cleaner. Thanks.

sbc100 commented on test/browser/test_glfw3_default_hints.c:
```diff
@@ -0,0 +1,86 @@
+#include <assert.h>
+#include <emscripten/html5.h>
+
+static void checkDefaultWindowHints()
```
Open brace on same line as func name.

ypujante replied:
Done

sbc100 commented on test/browser/test_glfw3_default_hints.c:
```diff
@@ -0,0 +1,86 @@
+
+  // reset hints
+  glfwDefaultWindowHints();
+  assert(EM_ASM_INT(return GLFW.hints[0x00021005];) == 24);
```
I wonder if you can test this without using `EM_JS` and instead using `glfwGetWindowParam`?

Then maybe we can avoiding hardcoded values like 0x00021005 in the test?

ypujante replied:
The issue is that `glfwGetWindowParam` is not part of the GLW3 api. According to the [documentation](https://www.glfw.org/docs/3.0/moving.html) it has been replaced by `glfwGetWindowAttrib`, but it requires a window as an argument which I don't have in the test because I never create a window.

sbc100 reviewed (commented):
lgtm with some comments

ypujante commented:
@sbc100 I have updated the PR according to your feedback.

Looking at the source code, I noticed [this](https://github.com/emscripten-core/emscripten/blob/d5f8412a4fdf9fc94b1ac85f43ddf689d3013abd/src/library_glfw.js#L54) as well

```javascript
this.attributes = GLFW.hints;
``` 

which is a similar "mistake" and should probably be fixed.

If you write a test code like this:

```javascript
glfwDefaultWindowHints();
glfwWindowHint(GLFW_DEPTH_BITS, 16);
window = glfwCreateWindow(...)
assert(glfwGetWindowAttrib(window, GLFW_DEPTH_BITS) == 16);
// this changes the window hints and due to the shallow copy, 
// also the attributes of the window just created...
glfwWindowHint(GLFW_DEPTH_BITS, 24); 
assert(glfwGetWindowAttrib(window, GLFW_DEPTH_BITS) == 16); // this will fail...
``` 

Do you want me to fix it part of this PR as well?

```javascript
// trivial fix...
this.attributes = Object.assign({}, GLFW.hints);
```

ypujante commented:
@sbc100 I ended up fixing the attributes assignment as well since it is such a trivial fix (and added a test for it)

I don't understand why so many tests failed and they seem to be unrelated to my changes. The test that I added passes ok:

```text
  test_glfw3_default_hints (test_browser.browser) ... cache:INFO: generating system asset: symbol_lists/bd12648a8db173908b13aa381a9bd5bab1858118.json... (this will be cached in "/root/cache/symbol_lists/bd12648a8db173908b13aa381a9bd5bab1858118.json" for subsequent builds)
cache:INFO:  - ok
ok (0.931s)
```

sbc100 commented:
Can you try merging (or rebasing) to get the latest changes from the main branch?

sbc100 reviewed (approved):
lgtm

ypujante commented:
There has been some progress, but still some tests failure which I believe are unrelated to my changes:

```
FAIL [121.921s]: test_idbstore_sync_jspi (test_browser.browser)
FAIL [122.945s]: test_idbstore_sync_worker (test_browser.browser)
FAIL [121.052s]: test_in_flight_memfile_request_O0 (test_browser.browser)
FAIL [121.069s]: test_in_flight_memfile_request_O1 (test_browser.browser)
FAIL [121.922s]: test_in_flight_memfile_request_O2 (test_browser.browser)
```

ypujante commented:
Everything is green! Yay!

sbc100 commented on test/test_browser.py:
```diff
@@ -2999,6 +2999,10 @@ def in_html(expected):
 
+  @requires_graphics_hardware
+  def test_glfw3_default_hints(self):
+      self.btest_exit('test_glfw3_default_hints.c', args=['-sUSE_GLFW=3', '-lglfw', '-lGL'])
```
We use 2-space indentation for python code

ypujante replied:
fixed

ypujante commented on test/browser/test_glfw3_default_hints.c:
```diff
@@ -0,0 +1,105 @@
+  int ok = EM_ASM_INT({
+    var res = 1;
+    for (var k in TEST_GLFW3_DEFAULTS_HINTS) {
+      if(GLFW.defaultHints[k] !== TEST_GLFW3_DEFAULTS_HINTS[k])
```
fixed

ypujante commented:
I fixed the small indentation issues and everything still green :)
